src/NHibernate/Async/AdoNet/ConnectionManager.cs

```diff
@@ -0,0 +1,69 @@
+﻿//------------------------------------------------------------------------------
+// <auto-generated>
+//     This code was generated by AsyncGenerator.
+//
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated.
+// </auto-generated>
+//------------------------------------------------------------------------------
+
+
+using System;
+using System.Data;
+using System.Data.Common;
+using System.Runtime.Serialization;
+using System.Security;
+
+using NHibernate.Engine;
+
+namespace NHibernate.AdoNet
+{
+	using System.Threading.Tasks;
+	using System.Threading;
+	/// <content>
+	/// Contains generated async methods
+	/// </content>
+	public partial class ConnectionManager : ISerializable, IDeserializationCallback
+	{
+
+		public async Task<DbConnection> GetConnectionAsync(CancellationToken cancellationToken)
+		{
+			cancellationToken.ThrowIfCancellationRequested();
+			if (connection == null)
+			{
+				if (ownConnection)
+				{
+					connection = await (Factory.ConnectionProvider.GetConnectionAsync(cancellationToken)).ConfigureAwait(false);
+					if (Factory.Statistics.IsStatisticsEnabled)
+					{
+						Factory.StatisticsImplementor.Connect();
+					}
+				}
+				else if (session.IsOpen)
+				{
+					throw new HibernateException("Session is currently disconnected");
+				}
+				else
+				{
+					throw new HibernateException("Session is closed");
```

Comment:
Here the problem is that the preconditions are not placed at the beginning of the method, that's why a tail method is not generated. By default, the generator will create a tail method if the method would be rewritten to:
```
public DbConnection GetConnection()
{
    if (connection == null && !ownConnection && session.IsOpen)
    {
        throw new HibernateException("Session is currently disconnected");
    }
    if (connection == null && !ownConnection)
    {
        throw new HibernateException("Session is closed");
    }
    if (connection != null)
    {
        return connection;
    }
    connection = Factory.ConnectionProvider.GetConnection();
    if (Factory.Statistics.IsStatisticsEnabled)
    {
        Factory.StatisticsImplementor.Connect();
    }
    return connection;
}
```
Another option is add a custom precondition checker ([example](https://github.com/maca88/AsyncGenerator/blob/master/Source/AsyncGenerator.Tests/Preconditions/Fixture.cs#L139)), but still the precodnitions must be moved before the async invocation.